### PR TITLE
added basic networking benchmarks

### PR DESCRIPTION
**Pull request by JanEggers**
## Proposed Changes

I felt there is a need to have a reproducable way to benchmark changes like #949 
This pr does not affect the production code and the benchmakrs are quite basic.
My intention is to provide more pullrequests that improve performance / reduce allocations and this pr should establish a common ground so I can prove that the changes actually benifit the performance.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality) 
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

### CODE REVIEW

File: projects/Benchmarks/Networking/RabbitMqBroker.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Threading;
+using Ductus.FluentDocker.Builders;
+
+namespace Benchmarks.Networking
+{
+    public class RabbitMqBroker
```
Comment (michaelklishin) [inline projects/Benchmarks/Networking/RabbitMqBroker.cs:7]: Can we make this say `RabbitMQ` and not `RabbitMq`?

File: projects/Benchmarks/Networking/RabbitMqBroker.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+            var broker = new Builder().UseContainer()
+               .UseImage("rabbitmq")
+               .ExposePort(5672, 5672)
+               .WaitForPort("5672/tcp", 40000 /*40s*/)
+               .ReuseIfExists()
+               .WithName("rabbitmq")
+               .Build()
+               .Start();
+
+            Thread.Sleep(5000); // the broker needs some more time after the port is open to be operational
```
Comment (michaelklishin) [inline projects/Benchmarks/Networking/RabbitMqBroker.cs:20]: This approach is fine but you can also run `rabbitmqctl await_startup` and then wait for no more than a second or so.

File: projects/Benchmarks/Networking/RabbitMqBroker.cs (hunk 1 above)
Comment (michaelklishin) [reply projects/Benchmarks/Networking/RabbitMqBroker.cs:7]: Or even `RabbitMQNode`.

**Comment by michaelklishin** [comment]
@JanEggers thank you. @stebet has a bunch of benchmarks/profiling tasks we use. It would be great to convert them to use the same "framework" as in this PR.

It also would be great to have documented a way to run these with some kind of output (could be HTML files or console output) from the command line and not from VS.

File: projects/Benchmarks/Networking/RabbitMqBroker.cs (hunk 1 above)
Comment (JanEggers) [reply projects/Benchmarks/Networking/RabbitMqBroker.cs:7]: sure

File: projects/Benchmarks/Networking/RabbitMqBroker.cs (hunk 2 above)
Comment (JanEggers) [reply projects/Benchmarks/Networking/RabbitMqBroker.cs:20]: Failed to execute rabbitmqctl await_startup error: Error: unable to perform an operation on node 'rabbit@ed1d67aa8b5a'. Please see diagnostics information and suggestions below.
Most common reasons for this are:
 * Target node is unreachable (e.g. due to hostname resolution, TCP connection or firewall issues)
 * CLI tool fails to authenticate with the server (e.g. due to CLI tool's Erlang cookie not matching that of the server)
 * Target node is not running
In addition to the diagnostics info below:
 * See the CLI, clustering and networking guides on https://rabbitmq.com/documentation.html to learn more
 * Consult server logs on node rabbit@ed1d67aa8b5a
 * If target node is configured to use long node names, don't forget to use --longnames with CLI tools
DIAGNOSTICS
===========
attempted to contact: [rabbit@ed1d67aa8b5a]
rabbit@ed1d67aa8b5a:
  * connected to epmd (port 4369) on ed1d67aa8b5a
  * epmd reports: node 'rabbit' not running at all
                  other nodes on ed1d67aa8b5a: ['rabbitmqcli-47-rabbit']
  * suggestion: start the node
Current node details:
 * node name: 'rabbitmqcli-87-rabbit@ed1d67aa8b5a'
 * effective user's home directory: /var/lib/rabbitmq
 * Erlang cookie hash: Ga7ZD0sbg8lx30HyB3nk/A==

Comment (JanEggers) [reply projects/Benchmarks/Networking/RabbitMqBroker.cs:20]: the command does not seem to work

**Comment by JanEggers** [comment]
@michaelklishin  FYI https://benchmarkdotnet.org/articles/configs/exporters.html

you have been useing benchmark.net before I just added new ones with docker so you can test without manually installing a broker.

File: projects/Benchmarks/Networking/RabbitMqBroker.cs (hunk 2 above)
Comment (JanEggers) [reply projects/Benchmarks/Networking/RabbitMqBroker.cs]: setting a nodename fixed the issue

File: projects/Benchmarks/Networking/Networking_BasicDeliver.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        {
+            _container = RabbitMQBroker.Start(); 
+        }
+
+        [GlobalCleanup]
+        public void GlobalCleanup()
+        {
+            _container.Dispose();
+        }
+
```
Comment (sungam3r) [inline projects/Benchmarks/Networking/Networking_BasicDeliver.cs:29]:
```suggestion
```

File: projects/Benchmarks/Networking/Networking_BasicDeliver.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                {
+                    if (Interlocked.Increment(ref consumed) == messageCount)
+                    {
+                        tcs.SetResult(true);
+                    }
+                };
+                model.BasicConsume(queue.QueueName, true, consumer);
+
+                const string publish1 = "hello world";
+                byte[] body = Encoding.UTF8.GetBytes(publish1);
```
Comment (sungam3r) [inline projects/Benchmarks/Networking/Networking_BasicDeliver.cs:59]: Worth to move into `GlobalSetup` ?

File: projects/Benchmarks/Networking/Networking_BasicDeliver_ExistingConnection.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+            var cf = new ConnectionFactory { ConsumerDispatchConcurrency = 2 };
+            _connection = cf.CreateConnection();
+        }
+
+        [GlobalCleanup]
+        public void GlobalCleanup()
+        {
+            _container.Dispose();
+            _connection.Dispose();
```
Comment (sungam3r) [inline projects/Benchmarks/Networking/Networking_BasicDeliver_ExistingConnection.cs:32]:
```suggestion
            _connection.Dispose();
            _container.Dispose();
```

File: projects/Benchmarks/Networking/Networking_BasicDeliver.cs (hunk 4 above)
Comment (JanEggers) [reply projects/Benchmarks/Networking/Networking_BasicDeliver.cs:59]: when we get to the unit of allocated bytes but as long we are at the range of MB it matters little where we allocate the 11 byte payload

File: projects/Benchmarks/Networking/Networking_BasicDeliver.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                for (int i = 0; i < messageCount; i++)
+                {
+                    var bp = model.CreateBasicProperties();
```
Comment (bollhals) [inline projects/Benchmarks/Networking/Networking_BasicDeliver.cs:62]: you can reuse the same properties or even pass null (I think)

File: projects/Benchmarks/Benchmarks.csproj [hunk 7]
```diff
@@ -9,6 +9,7 @@
 
   <ItemGroup>
     <PackageReference Include="BenchmarkDotNet" Version="0.12.1" />
+    <PackageReference Include="Ductus.FluentDocker" Version="2.7.3" />
```
Comment (bollhals) [inline projects/Benchmarks/Benchmarks.csproj:12]: Didn't know that lib, thanks =)

**Comment by bollhals** [comment]
>> @JanEggers thank you. @stebet has a bunch of benchmarks/profiling tasks we use. It would be great to convert them to use the same "framework" as in this PR.
>> 
>> It also would be great to have documented a way to run these with some kind of output (could be HTML files or console output) from the command line and not from VS.
> @michaelklishin FYI https://benchmarkdotnet.org/articles/configs/exporters.html
> 
> you have been useing benchmark.net before I just added new ones with docker so you can test without manually installing a broker.

You can also run them also outside of VS if needed. => https://benchmarkdotnet.org/articles/guides/how-to-run.html

Also while this benchmark will be unstable due to the networking involved, I still like it to get a sense as well as actual allocation numbers.

**Comment by JanEggers** [comment]
can someone of you guys please run the benchmark im having problems with the broker in the docker.
the command @michaelklishin pointed to me sometimes fails because the nodenames dont match.

when i start the container before running the benchmark it runns fine though

**Comment by bollhals** [comment]
> can someone of you guys please run the benchmark im having problems with the broker in the docker.
> the command @michaelklishin pointed to me sometimes fails because the nodenames dont match.
> 
> when i start the container before running the benchmark it runns fine though

Can this be fixed so this PR could be merged?

**Comment by JanEggers** [comment]
@bollhals I fixed the waiting code but I had to add some retry.

To me it seem that when executing the command the first time the broker did not yet pick up the node name resulting in the command to fail

**Comment by bollhals** [comment]
What's left here to be done so it could get merged? @michaelklishin 

Also is @lsfera preferable to @JanEggers solution?

File: projects/Benchmarks/Networking/RabbitMqBroker.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+               .WaitForPort("5672/tcp", 40000 /*40s*/)
+               .ReuseIfExists()
+               .WithName("rabbitmq")
+               .WithEnvironment("NODENAME=rabbit1")
+               .Build()
+               .Start();
+
+            for (int i = 0; i < 10; i++)
+            {
+                var response = broker.Execute("rabbitmqctl--node rabbit1 await_startup ");
```
Comment (michaelklishin) [inline projects/Benchmarks/Networking/RabbitMqBroker.cs:29]: There's a missing space between `rabbitmqctl` and `--node`

Comment (JanEggers) [reply projects/Benchmarks/Networking/RabbitMqBroker.cs]: thx @michaelklishin for catching that.  I rebased and pulled @lsfera changes as that probe looks cleaner to me

**Comment by michaelklishin** [comment]
I could not get the recommended `dotnet benchmark` tool to run (it seems to have a dependency that requires .NET Core 2.1, or at least I could not figure out what exactly did) so the way to run these from the command line is

```
dotnet run -c Release -f netcoreapp5.0 --project ./projects/Benchmarks/ --filter Networking_BasicDeliver_LongLivedConnection
```

**Comment by michaelklishin** [comment]
@JanEggers a6cafe3fd7d50f5c64e4634d3547b0a69a5a1dfa did not make it into master, unfortunately. Can you please re-submit?

**Comment by michaelklishin** [comment]
I merged this manually but without a few most recent commits by @JanEggers. No worries, we can always refine what we now have in master.

**Pull request closed without merging**